operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,422 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ciliumendpointbatch
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"k8s.io/client-go/util/workqueue"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// sequentialLetters contains lower case alphabets without vowels and few numbers.
+	// skipped vowels and numbers [0, 1] to avoid generating controversial names.
+	sequentialLetters = []rune("bcdfghjklmnpqrstvwxyz2456789")
+)
+
+// cebTracker holds the desired state of CiliumEndpointBatch and list of ceps to be removed
+// in next sync with k8s-apiserver.
+type cebTracker struct {
+	// Mutex to protect cep insert/removal in ceb and removedCeps
+	backendMutex lock.RWMutex
```

Comment:
This mutex has a comment stating that it is protecting the `ceb` and `removedCeps` variables, however there are a lot of places in this file where the variables are not protected with the mutex. For example [1], but there are other places.

Reply:
Comment made it clear `mutex` to protect `cep insert/removal` in CEB, because that is critical. we don't want to process `ceb.Endpoints` in the middle of update. 
However, I'm being conservative on using `lock` on various places, using `lock` on all places would hit our performance noticeably and i have skipped where it's not critical . Traditionally lock is avoided during `log` calls, because those are expensive api calls. 
If you prefer we need to use `lock` on all places, i'll be happy to update in code.

Reply:
if `ceb.Endpoints` is being protected by the mutex so should `len(cebTracker.ceb.Endpoints)` be also protected by the mutex which there are a lots of places where it's not protected by it.

Reply:
if `ceb.Endpoints` is being protected by the mutex so should `len(cebTracker.ceb.Endpoints)` be also protected by the mutex which there are a lots of places where it's not protected by it.

Reply:
okay, updated lock on couple of places. Not sure why you had mentioned `lot of places`